Refactor SetFields; allow overriding behavior when building auto types (#2875)
graphql-dotnet/graphql-dotnet: docs2/site/docs/migrations/migration5.md

serializing objects to and from `string` values. For the `GraphQL.SystemTextJson`
and `GraphQL.NewtonsoftJson` libraries, these serialize and deserialize to JSON strings.

### 6. `AutoRegisteringObjectGraphType` and `AutoRegisteringInputObjectGraphType` enhancements

#### Overridable base functionality

The classes can be overridden, providing the ability to customize behavior of automatically
generated graph types. For instance, to exclude properties marked with a custom attribute
called `[InternalUse]` you could write this:

```csharp
private class CustomAutoObjectType<T> : AutoRegisteringObjectGraphType<T>
{

>     protected override IEnumerable<FieldType> CreateFieldTypeList()

Reviewer: CreateFieldTypeList -> ProvideFields ?

Reviewer: I mean I don't like `list` in name a bit.

Author: Sure

    {
        var props = GetRegisteredProperties();
        foreach (var prop in props)
        {

>             if (prop.GetCustomAttribute<InternalUseAttribute>() != null)

Reviewer: ```suggestion
            if (prop.IsDefined(typeof(InternalUseAttribute)))
```

                yield return CreateFieldType(prop);
        }
    }
}
```

Similarly, by overriding `CreateFieldType` you can change the default name, description,
graph type, or other information applied to each generated field.

If you utilize dependency injection within your schema, you can register your custom graph

> type over top of the usual one as follows:

Reviewer: usual ?

Author: Umm, yes?

Reviewer: usual type?

Author: usual aka normal aka standard.....

Author: Do you like "standard" better?  There's probably a better way to word it, for sure.  I was trying to say that you can register a custom generic auto-registering graph type class.... to be instantiated instead of the.... i'm stumbling over my own words again.  You can see from the demo what I was trying to say.  Feel free to suggest a better sentence here.

Author: Maybe this?

```suggestion
type to be used instead of the standard type as follows:
```

Reviewer: standard -> built-in ? (if we are talking about autoregisteringgraphtype)

Author: ```suggestion
type to be used instead of the built-in type as follows:
```

```cs
services.AddSingleton(typeof(AutoRegisteringObjectGraphType<>), typeof(CustomAutoObjectType<>));
```

> Then any graph types defined as `AutoRegisteringObjectGraphType<...>` will use your custom

Reviewer: types -> type ?
